hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/crypto/CryptoInputStream.java

```diff
@@ -336,22 +338,50 @@ public int read(long position, byte[] buffer, int offset, int length)
       }
       
       return n;
-    } catch (ClassCastException e) {
+    } else {
       throw new UnsupportedOperationException("This stream does not support " +
           "positioned read.");
     }
   }
+
+  /**
+   * Positioned read using {@link ByteBuffer}s. This method is thread-safe.
+   */
+  @Override
+  public int read(long position, final ByteBuffer buf)
+      throws IOException {
+    checkStream();
+    if (in instanceof ByteBufferPositionedReadable) {
+      int bufPos = buf.position();
+      final int n = ((ByteBufferPositionedReadable) in).read(position, buf);
+      if (n > 0) {
+        // This operation does not change the current offset of the file
+        decrypt(position, buf, n, bufPos);
+      }
+
+      return n;
+    } else {
+      throw new UnsupportedOperationException("This stream does not support " +
+          "positioned reads with byte buffers.");
+    }
+  }
   
   /**
    * Decrypt length bytes in buffer starting at offset. Output is also put 
    * into buffer starting at offset. It is thread-safe.
    */
   private void decrypt(long position, byte[] buffer, int offset, int length) 
       throws IOException {
-    ByteBuffer inBuffer = getBuffer();
-    ByteBuffer outBuffer = getBuffer();
+    ByteBuffer inBuffer = null;
+    ByteBuffer outBuffer = null;
     Decryptor decryptor = null;
     try {
+      // TODO we should be able to avoid copying chunks between the input buf,
```

Comment:
I'd rather creating a new JIRA for this (now), over leaving a TODO in the code. You know that TODO is going to get forgotten about otherwise....

Reply:
Removed the TODO and filed HDFS-14417